pkg/user/tx_client.go

```diff
@@ -166,83 +170,124 @@ func SetupTxClient(
+	return client.BroadcastPayForBlobWithAccount(ctx, client.defaultAccount, blobs, opts...)
 }
 
-func (s *TxClient) BroadcastPayForBlobWithAccount(ctx context.Context, account string, blobs []*blob.Blob, opts ...TxOption) (*sdktypes.TxResponse, error) {
-	s.mtx.Lock()
-	defer s.mtx.Unlock()
-	if err := s.checkAccountLoaded(ctx, account); err != nil {
+func (client *TxClient) BroadcastPayForBlobWithAccount(ctx context.Context, account string, blobs []*blob.Blob, opts ...TxOption) (*sdktypes.TxResponse, error) {
+	client.mtx.Lock()
+	defer client.mtx.Unlock()
+	if err := client.checkAccountLoaded(ctx, account); err != nil {
 		return nil, err
 	}
 
-	txBytes, _, err := s.signer.CreatePayForBlobs(account, blobs, opts...)
+	blobSizes := make([]uint32, len(blobs))
+	for i, blob := range blobs {
+		blobSizes[i] = uint32(len(blob.Data))
+	}
+
+	gasLimit := uint64(float64(types.DefaultEstimateGas(blobSizes)) * client.gasMultiplier)
+	fee := uint64(math.Ceil(appconsts.DefaultMinGasPrice * float64(gasLimit)))
+	// prepend calculated params, so it can be overwritten in case the user has specified it.
+	opts = append([]TxOption{SetGasLimit(gasLimit), SetFee(fee)}, opts...)
+
+	txBytes, _, err := client.signer.CreatePayForBlobs(account, blobs, opts...)
 	if err != nil {
 		return nil, err
 	}
 
-	return s.broadcastTx(ctx, txBytes, account)
+	return client.broadcastTx(ctx, txBytes, account)
 }
 
 // SubmitTx forms a transaction from the provided messages, signs it, and submits it to the chain. TxOptions
 // may be provided to set the fee and gas limit.
-func (s *TxClient) SubmitTx(ctx context.Context, msgs []sdktypes.Msg, opts ...TxOption) (*sdktypes.TxResponse, error) {
-	resp, err := s.BroadcastTx(ctx, msgs, opts...)
+func (client *TxClient) SubmitTx(ctx context.Context, msgs []sdktypes.Msg, opts ...TxOption) (*sdktypes.TxResponse, error) {
+	resp, err := client.BroadcastTx(ctx, msgs, opts...)
 	if err != nil {
 		return resp, err
 	}
 
-	return s.ConfirmTx(ctx, resp.TxHash)
+	return client.ConfirmTx(ctx, resp.TxHash)
 }
 
-func (s *TxClient) BroadcastTx(ctx context.Context, msgs []sdktypes.Msg, opts ...TxOption) (*sdktypes.TxResponse, error) {
-	s.mtx.Lock()
-	defer s.mtx.Unlock()
-	account, err := s.getAccountNameFromMsgs(msgs)
+func (client *TxClient) BroadcastTx(ctx context.Context, msgs []sdktypes.Msg, opts ...TxOption) (*sdktypes.TxResponse, error) {
+	client.mtx.Lock()
+	defer client.mtx.Unlock()
+	account, err := client.getAccountNameFromMsgs(msgs)
 	if err != nil {
 		return nil, err
 	}
 
-	if err := s.checkAccountLoaded(ctx, account); err != nil {
+	if err := client.checkAccountLoaded(ctx, account); err != nil {
 		return nil, err
 	}
 
-	tx, account, _, err := s.signer.SignTx(msgs, opts...)
+	txBuilder, err := client.signer.txBuilder(msgs, opts...)
 	if err != nil {
 		return nil, err
 	}
 
-	txBytes, err := s.signer.EncodeTx(tx)
+	estimated := false
```

Comment:
```suggestion
	isFeeEstimated := false
```
For readability it may be better to call this feeEstimated (since there is also gas estimating)

Reply:
Actually this should be the other way around. Something like `hasUserSetFee` right?

Reply:
`hasUserSetFee` sgtm